runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/middleware/FlexibleChecksumsRequestMiddleware.kt

```diff
@@ -0,0 +1,117 @@
+import aws.smithy.kotlin.runtime.io.*
+import aws.smithy.kotlin.runtime.util.*
+import kotlinx.coroutines.*
+import kotlin.coroutines.coroutineContext
+
+/**
+ * Mutate a request to enable flexible checksums.
+ *
+ * If the checksum will be sent as a header, calculate the checksum.
+ *
+ * Otherwise, if it will be sent as a trailing header, calculate the checksum as asynchronously as the body is streamed.
+ * In this case, a [LazyAsyncValue] will be added to the execution context which allows the trailing checksum to be sent
+ * after the entire body has been streamed.
+ *
+ * @param checksumAlgorithmName the name of the algorithm used to calculate the checksum
+ */
+@InternalApi
+public class FlexibleChecksumsRequestMiddleware(private val checksumAlgorithmName: String) : ModifyRequestMiddleware {
+    public override suspend fun modifyRequest(req: SdkHttpRequest): SdkHttpRequest {
+        val logger = coroutineContext.getLogger<FlexibleChecksumsRequestMiddleware>()
+
+        val headerName = "x-amz-checksum-${checksumAlgorithmName.lowercase()}"
+        logger.debug { "Resolved header name: $headerName" }
+
+        // remove all checksum headers except for $headerName
+        // this handles the case where a user inputs a precalculated checksum, but it doesn't match the input checksum algorithm
+        req.subject.headers.removeAllChecksumHeadersExcept(headerName)
+
+        val checksumAlgorithm = checksumAlgorithmName.toHashFunction() ?: throw ClientException("Could not parse checksum algorithm $checksumAlgorithmName")
+
+        if (!checksumAlgorithm.isSupported) {
+            throw ClientException("Checksum algorithm $checksumAlgorithmName is not supported for flexible checksums")
+        }
+
+        if (req.subject.body.isEligibleForAwsChunkedStreaming) {
+            req.subject.header("x-amz-trailer", headerName)
+
+            val deferredChecksum = CompletableDeferred<String>()
+
+            if (req.subject.headers[headerName] != null) {
+                logger.debug { "User supplied a checksum, skipping asynchronous calculation" }
+
+                val checksum = req.subject.headers[headerName]!!
+                req.subject.headers.remove(headerName) // remove the checksum header because it will be sent as a trailing header
+
+                deferredChecksum.complete(checksum)
+            } else {
+                logger.debug { "Calculating checksum asynchronously" }
+
+                req.subject.body = req.subject.body
+                    .toHashingBody(checksumAlgorithm, req.subject.body.contentLength)
+                    .toCompletingBody(deferredChecksum, req.subject.body.contentLength)
+            }
+
+            req.subject.trailingHeaders.append(headerName, deferredChecksum)
+        } else if (req.subject.headers[headerName] == null) {
+            logger.debug { "Calculating checksum" }
+
+            val checksum = req.subject.body.readAll()?.hash(checksumAlgorithm)?.encodeBase64String()
```

Comment:
**correctness**: In addition to Ian's comment below about triggering when `contentLength` could be `null`. This is consuming the entire body without replacing it. This works for replayable streams only. Is that always true in this branch?

Reply:
No it's not, I'll need to handle that 👍